driver-reactive-streams/src/main/com/mongodb/reactivestreams/client/internal/gridfs/ResizingByteBufferFlux.java

```diff
@@ -51,19 +52,16 @@ public void subscribe(final CoreSubscriber<? super ByteBuffer> actual) {
             BaseSubscriber<ByteBuffer> subscriber = new BaseSubscriber<ByteBuffer>() {
                 private volatile ByteBuffer remainder;
                 private final AtomicLong requested = new AtomicLong();
-                private volatile boolean startedProcessing = false;
+                private final AtomicBoolean startedProcessing = new AtomicBoolean();
                 private volatile boolean finished = false;
 
                 @Override
                 protected void hookOnSubscribe(final Subscription subscription) {
                     sink.onCancel(() -> upstream().cancel());
                     sink.onRequest(l -> {
-                        synchronized (this) {
-                            requested.addAndGet(l);
-                            if (!startedProcessing) {
-                                startedProcessing = true;
-                                upstream().request(1);
-                            }
+                        requested.addAndGet(l);
+                        if (startedProcessing.compareAndSet(false, true)) {
+                            upstream().request(1);
                         }
                     });
```

Comment:
I am not sure synchronization is needed here at all, because of the following rules:

- [`Subscription` #1](https://github.com/reactive-streams/reactive-streams-jvm#user-content-3.1): _`Subscription.request` and `Subscription.cancel` MUST only be called inside of its `Subscriber` context._
- [`Subscriber` #7](https://github.com/reactive-streams/reactive-streams-jvm#user-content-2.7): _A `Subscriber` MUST ensure that all calls on its `Subscription`'s request and cancel methods are performed serially._

But, because I am not sure it is not needed (I may be missing or misunderstanding something), I preserved the synchronization.

Reply:
Its not related to `cancel`. 
Its to ensure that multiple calls to `subscription.request(n);` don't both trigger an upstream request.

Reply:
> Its not related to `cancel`.

Right, I simply quoted the two relevant rules in full.

> Its to ensure that multiple calls to subscription.request(n); don't both trigger an upstream request.

But according to the quoted rules, `request` cannot run concurrently with itself (nor can it run concurrently with `cancel`). This is what makes original `synchronization` and the `AtomicBoolean` replacing it, weird. This also makes using `AtomicLong` for `requested` weird: "at worst", it could be just `volatile`, and "at best" it may also need no synchronization at all, but I did analyze this part because it is irrelevant to the PR.

Reply:
Good spot, so this may never have been an issue or is no longer be an issue since we bumped the reactor version (2022.0.0  had some changes that fix compliance with the Reactive streams spec.)

Added JAVA-5114 to track